File: ios/APEXNative/APEX/Core/Engine/ExerciseLogging.swift
import Foundation

enum ExerciseLoggingKind: String, CaseIterable, Sendable {
    case strength
    case bodyweight
    case isometric
    case carry
    case cardio
    case mobility
    case interval
    case circuit
}

enum ExerciseLoggingField: String, Sendable {
    case reps
    case signedLoad
    case rir
    case duration
    case distance
    case contacts
    case completion
    case rounds
    case work
    case recovery
}

struct ExerciseLoggingDescriptor: Equatable, Sendable {
    let kind: ExerciseLoggingKind
    let fields: [ExerciseLoggingField]
    let isSupported: Bool
}

enum ExerciseProgress: Equatable, Sendable {
    case improved
    case maintained
    case regressed
    case adherence
    case incomparable
}

enum ExerciseLogging {
    static func descriptor(
        movementNamed name: String,
        movementID: String? = nil
    ) -> ExerciseLoggingDescriptor {
        descriptor(for: MovementTiming.movement(named: name, movementID: movementID))
    }

    static func descriptor(for movement: MovementTiming.Movement?) -> ExerciseLoggingDescriptor {
        let kind = kind(for: movement)
        switch kind {
        case .strength, .bodyweight:
            if movement?.prescriptionMode == "contacts" {
                return ExerciseLoggingDescriptor(
                    kind: kind,
                    fields: [.contacts],
                    isSupported: true
                )
            }
            return ExerciseLoggingDescriptor(
                kind: kind,
                fields: [.reps, .signedLoad, .rir],
                isSupported: true
            )
        case .isometric:
            return ExerciseLoggingDescriptor(
                kind: kind,
                fields: [.duration, .signedLoad],
                isSupported: true
            )
        case .carry:
            return ExerciseLoggingDescriptor(
                kind: kind,
                fields: [.duration, .distance, .signedLoad],
                isSupported: true
            )
        case .cardio:
            return ExerciseLoggingDescriptor(
                kind: kind,
                fields: [.duration, .distance],
                isSupported: true
            )
        case .mobility:
            return ExerciseLoggingDescriptor(
                kind: kind,
                fields: [.duration, .completion],
                isSupported: true
            )
        case .interval:
            return ExerciseLoggingDescriptor(
                kind: kind,
                fields: [.rounds, .work, .recovery],
                isSupported: true
            )
        case .circuit:
            return ExerciseLoggingDescriptor(kind: kind, fields: [], isSupported: false)
        }
    }

    static func derivedPaceSecondsPerKilometre(
        distanceMeters: Double,
        durationSeconds: Int
    ) -> Double? {
        guard distanceMeters > 0, durationSeconds > 0 else { return nil }
        return Double(durationSeconds) * 1_000 / distanceMeters
    }

    /// Clear irrelevant observations and make unassisted bodyweight an explicit zero.
    static func normalized(_ input: WorkoutSetInput) -> WorkoutSetInput {
        let descriptor = descriptor(
            movementNamed: input.exerciseName,
            movementID: input.movementID
        )
        return normalized(input, descriptor: descriptor)
    }

    static func normalized(
        _ input: WorkoutSetInput,
        descriptor: ExerciseLoggingDescriptor
    ) -> WorkoutSetInput {
        let clear = input.skipped || !descriptor.isSupported
        let has: (ExerciseLoggingField) -> Bool = { descriptor.fields.contains($0) }
        let load: Double? = {
            guard !clear, has(.signedLoad) else { return nil }
            if let weight = input.weightKG, weight.isFinite { return weight }
            return descriptor.kind == .bodyweight ? 0 : nil
        }()
        let effort: Int? = {
            guard !clear, has(.rir), let rir = input.rir, (0...5).contains(rir) else { return nil }
            return rir
        }()
        return WorkoutSetInput(
            exerciseID: input.exerciseID,
            exerciseName: input.exerciseName,
            setNumber: input.setNumber,
            weightKG: load,
            reps: !clear && has(.reps) ? positive(input.reps) : nil,
            rir: effort,
            movementID: input.movementID,
            durationSeconds: !clear && has(.duration) ? positive(input.durationSeconds) : nil,
            distanceMeters: !clear && has(.distance) ? positive(input.distanceMeters) : nil,
            contacts: !clear && has(.contacts) ? positive(input.contacts) : nil,
            rounds: !clear && has(.rounds) ? positive(input.rounds) : nil,
            workSeconds: !clear && has(.work) ? positive(input.workSeconds) : nil,
            recoverySeconds: !clear && has(.recovery) ? positive(input.recoverySeconds) : nil,
            skipped: input.skipped
        )
    }

    /// A completed row is valid only when its kind's measured facts are complete.
    static func isValid(_ input: WorkoutSetInput) -> Bool {
        let descriptor = descriptor(
            movementNamed: input.exerciseName,
            movementID: input.movementID
        )
        return isValid(input, descriptor: descriptor)
    }

    static func isResolved(_ input: WorkoutSetInput) -> Bool {
        input.skipped || isValid(input)
    }

    static func resolvingAsSkipped(_ input: WorkoutSetInput) -> WorkoutSetInput {
        var skipped = input
        skipped.skipped = true
        return normalized(skipped)
    }

    static func isValid(
        _ input: WorkoutSetInput,
        descriptor: ExerciseLoggingDescriptor
    ) -> Bool {
        guard descriptor.isSupported else { return false }
        let value = normalized(input, descriptor: descriptor)
        switch descriptor.kind {
        case .strength, .bodyweight:
            return descriptor.fields.contains(.contacts)
                ? value.contacts != nil
                : value.reps != nil
        case .isometric:
            return value.durationSeconds != nil
        case .carry:
            return (value.durationSeconds != nil) != (value.distanceMeters != nil)
        case .cardio:
            return value.durationSeconds != nil && value.distanceMeters != nil
        case .mobility:
            return true
        case .interval:
            return value.rounds != nil && value.workSeconds != nil && value.recoverySeconds != nil
        case .circuit:
            return false
        }
    }

    static func hasPrimaryFacts(
        descriptor: ExerciseLoggingDescriptor,
        reps: Int?,
        durationSeconds: Int?,
        distanceMeters: Double?,
        contacts: Int?,
        rounds: Int?,
        workSeconds: Int?,
        recoverySeconds: Int? = nil
    ) -> Bool {
        guard descriptor.isSupported else { return false }
        if descriptor.fields.contains(.completion) { return true }
        if descriptor.fields.contains(.contacts) { return (contacts ?? 0) > 0 }
        if descriptor.fields.contains(.rounds) || descriptor.fields.contains(.work) {
            return (rounds ?? 0) > 0 && (workSeconds ?? 0) > 0 && (recoverySeconds ?? 0) > 0
        }
        if descriptor.fields.contains(.reps) { return (reps ?? 0) > 0 }
        if descriptor.fields.contains(.duration) || descriptor.fields.contains(.distance) {
            return (durationSeconds ?? 0) > 0 || (distanceMeters ?? 0) > 0
        }
        return false
    }

    static func factSummary(_ log: WorkoutLog) -> [String] {
        if log.skipped { return ["Not completed"] }
        let descriptor = descriptor(
            movementNamed: log.exerciseName,
            movementID: log.movementID
        )
        var facts: [String] = []
        if descriptor.fields.contains(.reps), let reps = log.reps { facts.append("\(reps) reps") }
        if descriptor.fields.contains(.signedLoad), let load = log.weightKG {
            facts.append("\(number(load)) kg")
        }
        if descriptor.fields.contains(.rir), let rir = log.rir { facts.append("RIR \(rir)") }
        if descriptor.fields.contains(.duration), let duration = log.durationSeconds {
            facts.append("\(duration) sec")
        }
        if descriptor.fields.contains(.distance), let distance = log.distanceMeters {
            facts.append("\(number(distance)) m")
        }
        if descriptor.fields.contains(.contacts), let contacts = log.contacts {
            facts.append("\(contacts) contacts")
        }
        if descriptor.fields.contains(.rounds), let rounds = log.rounds {
            facts.append("\(rounds) rounds")
        }
        if descriptor.fields.contains(.work), let work = log.workSeconds {
            facts.append("\(work) sec work")
        }
        if descriptor.fields.contains(.recovery), let recovery = log.recoverySeconds {
            facts.append("\(recovery) sec recovery")
        }
        if descriptor.fields.contains(.completion), facts.isEmpty { facts.append("Completed") }
        if descriptor.kind == .cardio,
           let distance = log.distanceMeters,
           let duration = log.durationSeconds,
           let pace = derivedPaceSecondsPerKilometre(
            distanceMeters: distance,
            durationSeconds: duration
           ) {
            let rounded = Int(pace.rounded())
            facts.append(String(format: "%d:%02d /km", rounded / 60, rounded % 60))
        }
        return facts
    }

    private static func number(_ value: Double) -> String {
        value == value.rounded() ? String(Int(value)) : String(format: "%.1f", value)
    }

    private static func positive(_ value: Int?) -> Int? {
        guard let value, value > 0 else { return nil }
        return value
    }

    private static func positive(_ value: Double?) -> Double? {
        guard let value, value.isFinite, value > 0 else { return nil }
        return value
    }

    private static func kind(for movement: MovementTiming.Movement?) -> ExerciseLoggingKind {
        guard let movement else { return .strength }
        switch movement.entityType {
        case "cardio_modality":
            return .cardio
        case "movement_sequence":
            return .circuit
        case "conditioning_complex":
            return .interval
        case "balance_drill", "mobility_drill", "skill_drill", "yoga_pose":
            return .mobility
        case "resistance_isometric":
            return movement.prescriptionMode == "carry" ? .carry : .isometric
        case "plyometric":
            return .bodyweight
        case "resistance_dynamic":
            let bodyweight = !movement.loadable
                || movement.disciplines.contains("calisthenics")
                || movement.id == "assisted_pull_up_machine"
            return bodyweight ? .bodyweight : .strength
        default:
            return .strength
        }
    }
}

File: src/lib/exerciseLogging.ts
import {
  CARDIO_MODALITIES,
  CARDIO_ALIASES,
  MOVEMENTS,
  MOVEMENT_ALIASES,
  type Movement,
} from '../data/movements.ts'
import { resolveMovement } from './liftingTempo.ts'
import type { WorkoutLog } from './types.ts'

export type ExerciseLoggingKind =
  | 'strength'
  | 'bodyweight'
  | 'isometric'
  | 'carry'
  | 'cardio'
  | 'mobility'
  | 'interval'
  | 'circuit'

export type ExerciseLoggingField =
  | 'reps'
  | 'signedLoad'
  | 'rir'
  | 'duration'
  | 'distance'
  | 'contacts'
  | 'completion'
  | 'rounds'
  | 'work'
  | 'recovery'

export interface ExerciseLoggingDescriptor {
  kind: ExerciseLoggingKind
  fields: ExerciseLoggingField[]
  supported: boolean
  movementId: string | null
}

export interface ExerciseLoggingFacts {
  weight: number | null
  reps: number | null
  rir: number | null
  durationSeconds?: number | null
  distanceMeters?: number | null
  contacts?: number | null
  rounds?: number | null
  workSeconds?: number | null
  recoverySeconds?: number | null
}

const MOVEMENT_BY_ID = new Map(MOVEMENTS.map((movement) => [movement.id, movement]))
const CARDIO_BY_ID = new Map(CARDIO_MODALITIES.map((modality) => [modality.id, modality]))

function normalized(value: string): string {
  return value
    .replace(/\([^)]*\)/g, '')
    .normalize('NFD')
    .replace(/\p{Diacritic}/gu, '')
    .toLocaleLowerCase('en')
    .replace(/[^\p{Letter}\p{Number}]+/gu, '')
}

const CARDIO_BY_NAME = new Map(CARDIO_MODALITIES.map((modality) => [normalized(modality.name), modality]))
const MOVEMENT_ALIAS_BY_NAME = new Map(Object.entries(MOVEMENT_ALIASES).map(([name, movementID]) => [
  normalized(name),
  MOVEMENT_BY_ID.get(movementID) ?? null,
]))
const CARDIO_ALIAS_BY_NAME = new Map(Object.entries(CARDIO_ALIASES).map(([name, value]) => [
  normalized(name),
  CARDIO_BY_ID.get(value.modality) ?? null,
]))

function kindForMovement(movement: Movement): ExerciseLoggingKind {
  switch (movement.entityType) {
    case 'movement_sequence': return 'circuit'
    case 'conditioning_complex': return 'interval'
    case 'balance_drill':
    case 'mobility_drill':
    case 'skill_drill':
    case 'yoga_pose': return 'mobility'
    case 'resistance_isometric': return movement.prescriptionMode === 'carry' ? 'carry' : 'isometric'
    case 'plyometric': return 'bodyweight'
    case 'resistance_dynamic':
      return !movement.loadable
        || movement.disciplines.includes('calisthenics')
        || movement.id === 'assisted_pull_up_machine'
        ? 'bodyweight'
        : 'strength'
    default: return 'strength'
  }
}

function fieldsFor(kind: ExerciseLoggingKind, movement: Movement | null): ExerciseLoggingField[] {
  if (movement?.prescriptionMode === 'contacts') return ['contacts']
  switch (kind) {
    case 'strength':
    case 'bodyweight': return ['reps', 'signedLoad', 'rir']
    case 'isometric': return ['duration', 'signedLoad']
    case 'carry': return ['duration', 'distance', 'signedLoad']
    case 'cardio': return ['duration', 'distance']
    case 'mobility': return ['duration', 'completion']
    case 'interval': return ['rounds', 'work', 'recovery']
    case 'circuit': return []
  }
}

/** Resolve on every read. Neither the row nor the database stores a logging kind. */
export function descriptorForExercise(exercise: {
  name: string
  movement_id?: string | null
}): ExerciseLoggingDescriptor {
  const movement = exercise.movement_id ? MOVEMENT_BY_ID.get(exercise.movement_id) ?? null : null
  const cardio = exercise.movement_id ? CARDIO_BY_ID.get(exercise.movement_id) ?? null : null
  const resolvedMovement = movement
    ?? MOVEMENT_ALIAS_BY_NAME.get(normalized(exercise.name))
    ?? resolveMovement(exercise.name, MOVEMENTS, MOVEMENT_ALIASES)
  const resolvedCardio = cardio
    ?? CARDIO_BY_NAME.get(normalized(exercise.name))
    ?? CARDIO_ALIAS_BY_NAME.get(normalized(exercise.name))
    ?? null

  if (resolvedCardio) {
    return { kind: 'cardio', fields: fieldsFor('cardio', null), supported: true, movementId: resolvedCardio.id }
  }
  const kind = resolvedMovement ? kindForMovement(resolvedMovement) : 'strength'
  return {
    kind,
    fields: fieldsFor(kind, resolvedMovement),
    supported: kind !== 'circuit',
    movementId: resolvedMovement?.id ?? exercise.movement_id ?? null,
  }
}

const emptyFacts = (): Required<ExerciseLoggingFacts> => ({
  weight: null,
  reps: null,
  rir: null,
  durationSeconds: null,
  distanceMeters: null,
  contacts: null,
  rounds: null,
  workSeconds: null,
  recoverySeconds: null,
})

function positiveOrNull(value: number | null | undefined): number | null {
  return value != null && Number.isFinite(value) && value > 0 ? value : null
}

/** Clear irrelevant observations and make unassisted bodyweight an explicit zero. */
export function normalizeExerciseFacts(
  facts: ExerciseLoggingFacts,
  descriptor: ExerciseLoggingDescriptor,
  skipped: boolean,
): Required<ExerciseLoggingFacts> {
  if (skipped || !descriptor.supported) return emptyFacts()
  const has = (field: ExerciseLoggingField): boolean => descriptor.fields.includes(field)
  const load = has('signedLoad') && facts.weight != null && Number.isFinite(facts.weight)
    ? facts.weight
    : has('signedLoad') && descriptor.kind === 'bodyweight' ? 0 : null
  const rir = has('rir') && facts.rir != null && Number.isFinite(facts.rir)
    && facts.rir >= 0 && facts.rir <= 5 ? Math.round(facts.rir) : null
  return {
    weight: load,
    reps: has('reps') ? positiveOrNull(facts.reps) : null,
    rir,
    durationSeconds: has('duration') ? positiveOrNull(facts.durationSeconds) : null,
    distanceMeters: has('distance') ? positiveOrNull(facts.distanceMeters) : null,
    contacts: has('contacts') ? positiveOrNull(facts.contacts) : null,
    rounds: has('rounds') ? positiveOrNull(facts.rounds) : null,
    workSeconds: has('work') ? positiveOrNull(facts.workSeconds) : null,
    recoverySeconds: has('recovery') ? positiveOrNull(facts.recoverySeconds) : null,
  }
}

/** A completed row is valid only when its kind's measured facts are complete. */
export function isValidExerciseFacts(
  facts: ExerciseLoggingFacts,
  descriptor: ExerciseLoggingDescriptor,
): boolean {
  if (!descriptor.supported) return false
  const value = normalizeExerciseFacts(facts, descriptor, false)
  switch (descriptor.kind) {
    case 'strength':
    case 'bodyweight':
      return descriptor.fields.includes('contacts')
        ? value.contacts != null
        : value.reps != null
    case 'isometric': return value.durationSeconds != null
    case 'carry': return (value.durationSeconds != null) !== (value.distanceMeters != null)
    case 'cardio': return value.durationSeconds != null && value.distanceMeters != null
    case 'mobility': return true
    case 'interval': return value.rounds != null && value.workSeconds != null && value.recoverySeconds != null
    case 'circuit': return false
  }
}

export function derivePaceSecondsPerKilometre(
  distanceMeters: number,
  durationSeconds: number,
): number | null {
  if (!(distanceMeters > 0) || !(durationSeconds > 0)) return null
  return durationSeconds * 1_000 / distanceMeters
}

export function workoutLogFactSummary(log: WorkoutLog): string[] {
  const descriptor = descriptorForExercise({ name: log.exercise_name, movement_id: log.movement_id })
  if (log.skipped) return ['Not completed']
  const facts: string[] = []
  if (descriptor.fields.includes('reps') && log.reps != null) facts.push(`${log.reps} reps`)
  if (descriptor.fields.includes('signedLoad') && log.weight_kg != null) facts.push(`${log.weight_kg} kg`)
  if (descriptor.fields.includes('rir') && log.rir != null) facts.push(`RIR ${log.rir}`)
  if (descriptor.fields.includes('duration') && log.duration_seconds != null) facts.push(`${log.duration_seconds} sec`)
  if (descriptor.fields.includes('distance') && log.distance_meters != null) facts.push(`${log.distance_meters} m`)
  if (descriptor.fields.includes('contacts') && log.contacts != null) facts.push(`${log.contacts} contacts`)
  if (descriptor.fields.includes('rounds') && log.rounds != null) facts.push(`${log.rounds} rounds`)
  if (descriptor.fields.includes('work') && log.work_seconds != null) facts.push(`${log.work_seconds} sec work`)
  if (descriptor.fields.includes('recovery') && log.recovery_seconds != null) facts.push(`${log.recovery_seconds} sec recovery`)
  if (descriptor.fields.includes('completion') && facts.length === 0) facts.push('Completed')
  if (descriptor.kind === 'cardio' && log.distance_meters != null && log.duration_seconds != null) {
    const pace = derivePaceSecondsPerKilometre(log.distance_meters, log.duration_seconds)
    if (pace != null) {
      const rounded = Math.round(pace)
      facts.push(`${Math.floor(rounded / 60)}:${String(rounded % 60).padStart(2, '0')} /km`)
    }
  }
  return facts
}

/** Positive external kg × reps, including added load on bodyweight movements. */
export function loadedStrengthVolume(logs: WorkoutLog[]): number {
  return logs.reduce((total, log) => {
    const descriptor = descriptorForExercise({ name: log.exercise_name, movement_id: log.movement_id })
    if (log.skipped || (descriptor.kind !== 'strength' && descriptor.kind !== 'bodyweight')) return total
    return total + Math.max(0, log.weight_kg ?? 0) * Math.max(0, log.reps ?? 0)
  }, 0)
}
